### ISSUE TITLE
Ensure settings sync is resilient to server cache inconsistencies

### ISSUE THREAD

**Issue by Tyriar** [microsoft/vscode#90624]
Right now the cache is the source of truth on the server but there is an extreme edge case where it could become stale. Hopefully it can be fixed on the server-side but I want to make sure if this ever does happen that the client is resilient to this and can self-repair rather than going into a request loop that never fixes itself.

Consider the following scenario:

- Machine A: User updates settings
- Machine A: Request to write settings is made to the server where the cache becomes stale (database and cache do not match)
- Machine B: Does a sync but it doesn't take into account the change Machine A made since the cache is stale
- Machine B: User updates a settings, triggering a new request to write latest settings
- Machine B: Server returns 412 because it doesn't know about the actual latest settings record

Now the important part: Machine B must request to get the new latest settings **without providing an `if-none-match` header**. This will ensure that the cache is skipped and the latest version is fetched directly from the database. After this the new write should succeed and the cache will be repaired.

TL;DR: If a write fails because of a 412, the following request for the latest must not include `if-none-match`.

**Comment by Tyriar** [comment]
Ideally Feb as it should be simple

**Comment by sandy081** [comment]
May I know why the writes from B go in a loop? Are you saying that server cache will not get updated at all with the write from A?

**Comment by Tyriar** [comment]
MB asks server for write, server rejects with 412 (only db is validated for latest)
MB asks server for latest, server says everything is fine with 304 (cache is used if it exists)

**Comment by sandy081** [comment]
> MB asks server for latest, server says everything is fine with 304 (cache is used if it exists)
MB asks with the `lastSyncRef` as `If-None-Match` header which is not the latest on the server because MA has written. So, server should not return `304` right?

**Comment by Tyriar** [comment]
No, it would be 304 because the ask for latest went to the cache and the write went to the db and they differed. This scenario is to work around if the server gets into state where the cache has stale data, for example if the cache is offline for a few minutes. These things would self heal on the server after 6 minutes anyway as the cache item would expire, also this sort of thing will likely rarely if ever happen but I'm just trying to cover all bases. The self-healing would not occur if regular polling continues or a request loop happens.

**Comment by sandy081** [comment]
I am not comfortable to introduce this debt in client where the server is not returning what it is expected. 

> also this sort of thing will likely rarely if ever happen but I'm just trying to cover all bases

Introducing such a debt is hard to change in future. So not sure if its a good idea to do this on speculation. It would be good to know why it happened (when it happened) and figure out the root cause and fix it ?

**Comment by Tyriar** [comment]
Could we add one of the request assertion tests for this too?

**Comment by sandy081** [comment]
Sure

**Comment by sandy081** [comment]
@Tyriar Not sure how to verify this during testing.

**Comment by sandy081** [comment]
I would suggest @Tyriar to verify

**Comment by Tyriar** [comment]
Verified via review, it's pretty difficult to verify this properly. If all syncs go via `AbstractSynchroniser.sync` then it looks good.
